lib/charms/nginx_ingress_integrator/v0/ingress.py

```diff
@@ -189,40 +249,76 @@ def _on_relation_changed(self, event):
             )
             self.model.unit.status = BlockedStatus(
-                "Missing fields for ingress: {}".format(", ".join(missing_fields))
+                f"Missing fields for ingress: {', '.join(missing_fields)}"
             )
 
+        # Conform to charm-relation-interfaces.
+        if "name" in ingress_data and "port" in ingress_data:
+            name = ingress_data["name"]
+            port = ingress_data["port"]
+        else:
+            name = ingress_data["service-name"]
+            port = ingress_data["service-port"]
+        event.relation.data[self.model.app]["url"] = f"http://{name}:{port}/"
+
         # Create an event that our charm can use to decide it's okay to
         # configure the ingress.
         self.charm.on.ingress_available.emit()
 
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle a relation-broken event in the ingress relation."""
+        if not self.model.unit.is_leader():
+            return
 
-class IngressProvides(IngressBaseProvides):
-    """This class defines the functionality for the 'provides' side of the 'ingress' relation.
+        # Create an event that our charm can use to remove the ingress resource.
+        self.charm.on.ingress_broken.emit(event.relation)
 
-    Hook events observed:
-        - relation-changed
-    """
+class IngressProxyProvides(Object):
 
     def __init__(self, charm):
-        super().__init__(charm, "ingress")
+        super().__init__(charm, "ingress-proxy")
         # Observe the relation-changed hook event and bind
         # self.on_relation_changed() to handle the event.
-        self.framework.observe(charm.on["ingress"].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on["ingress-proxy"].relation_changed, self._on_relation_changed)
         self.charm = charm
 
+    def _on_relation_changed(self, event):
+        """Handle a change to the ingress relation.
 
-class IngressProxyProvides(IngressBaseProvides):
+        Confirm we have the fields we expect to receive."""
+        # `self.unit` isn't available here, so use `self.model.unit`.
+        if not self.model.unit.is_leader():
+            return
+        
+        if not event.relation.data[event.app]:
+            LOGGER.info("Ingress-proxy hasn't finished configuring, waiting until relation is changed again.")
+            return
 
-    def __init__(self, charm):
-        super().__init__(charm, "ingress")
-        # Observe the relation-changed hook event and bind
-        # self.on_relation_changed() to handle the event.
-        self.framework.observe(charm.on["ingress"].relation_changed, self._on_relation_changed)
-        self.charm = charm
+        ingress_data = {
+            field: event.relation.data[event.app].get(field)
+            for field in REQUIRED_INGRESS_RELATION_FIELDS | OPTIONAL_INGRESS_RELATION_FIELDS
+        }
+
+        missing_fields = sorted(
+            [
```

Comment:
Consider making this a generator since I sort will re-create the list anyway, that way we avoid defining a list and then immediately throwing it away